src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonPolymorphicAttribute.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// When placed on a type, indicates that the type should be serialized polymorphically.
+    /// </summary>
+    [System.Runtime.Versioning.RequiresPreviewFeatures("Polymorphic serialization is currently in preview.")]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = false, Inherited = false)]
+    public sealed class JsonPolymorphicAttribute : JsonAttribute
+    {
+        /// <summary>
+        /// Gets or sets a custom type discriminator property name for the polymorhic type.
+        /// Uses the default '$type' property name if left unset.
+        /// </summary>
+        public string? CustomTypeDiscriminatorPropertyName { get; set; }
```

Comment:
As a future feature that is often used in "entity" models, support specifying the discriminator metadata coming from an actual property on a base class typically an integer- or string-based property. That way there is no special "$" metadata embedded in the JSON.

Reply:
Could you give an example? As it stands this feature lets you change a polymorphic object from `{ "$type" : "x", ... }` to `{ "id" : "x", ... }` (but the property needs to be at the start of the JSON object). We can relax deserialization so that integer properties are tolerated, but I'm refraining from doing that before we get user requests for it.

Reply:
I'm thinking the scenario(s) Steve is referencing might require a way to specify the metadata property name dynamically/at runtime, in addition to this model where it is specified at design time.

Reply:
That should be in `JsonPolymorphismConfiguration.cs`. Plan is to eventually make that a settable property in `JsonTypeInfo`.